packages/wp-story-editor/src/index.js

```diff
@@ -58,10 +63,20 @@ const initialize = (id, settings, flags) => {
 
   initializeTracking('Editor');
 
+  // @todo Use more descriptive key names for mediaPickers.
+  const config = {
+    ...apiConfig,
+    mediaPickers: {
+      caption: useCaptionMediaPicker,
+      library: useLibraryMediaPicker,
+      form: useFormMediaPicker,
+    },
```

Comment:
Hmm I don't think we should separate media pickers like that, as this couples feature development in `story-editor` to changes in `wp-story-editor`. So for every new media feature I want to add, I need to edit both packages. It also causes lots of repetition because one has create new pickers over and over again.

For example, when I want to add a new "Background Audio" feature in `story-editor` (which works similar to captions), I must first create a new `useBackgroundAudioMediaPicker` in `wp-story-editor`, which is not ideal.

Is there another way we can achieve similar media integration with less coupling?

I see that Gutenberg has implemented `wp.media` integration with a custom `MediaUpload` component that's being added via a `editor.MediaUpload` filter:

https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/edit-post/src/hooks/components/index.js

https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/media-utils/src/components/media-upload/index.js

But we would need a solution without class components or filters...

Any ideas on how to make that possible? Lets' discuss!

Reply:
That's a valid point, let me try another approach.

Reply:
Without any hook/filter, the other way I can think of doing this would be to pass the `MediaUpload` component once, via the configuration like this 
https://github.com/rtCamp/web-stories-wp/blob/f49ea36df9356a5753d302df03331c5973054f32/packages/wp-story-editor/src/index.js#L62-L66

And we document, what props and their callback arguments we are expecting from `MediaUpload` as documented by Gutenberg here https://github.com/WordPress/gutenberg/tree/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/media-upload

and then use it like this
https://github.com/rtCamp/web-stories-wp/blob/f49ea36df9356a5753d302df03331c5973054f32/packages/story-editor/src/components/library/panes/media/local/mediaPane.js#L365-L382

Will that make sense?
Updated the PR with those changes.

Reply:
But even in that case,  some callbacks are using WordPress-specific code, for example https://github.com/rtCamp/web-stories-wp/blob/f49ea36df9356a5753d302df03331c5973054f32/packages/story-editor/src/components/library/panes/media/local/mediaPane.js#L222-L223 

Probably the `MediaUpload` component's code needs more improvement so that the callback arguments are more platform agnostic.

Reply:
I'll go through the code now.

At first glance, passing such a component with a `render` render prop sounds reasonable 👍 

> But even in that case, some callbacks are using WordPress-specific code, for example

Can we move the `getResourceFromMediaPicker` call to `useMediaPicker`? It shouldn't be `mediaPane.js`'s job to get a resource.

Basically, in places like this:

https://github.com/rtCamp/web-stories-wp/blob/fb7cfcb6d48ced117c908dca2d7743ef7ad60e68/packages/wp-story-editor/src/components/mediaUpload/useMediaPicker/useMediaPicker.js#L216-L233

We would do `onSelect(getResourceFromMediaPicker(attachment))`.